File: src/flutter_xattr.c
#ifdef linux

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <sys/xattr.h>
#include "flutter_xattr.h"

// get an attribute on a file
FFI_PLUGIN_EXPORT char *get_attribute(char *path, char *attribute)
{
    char value[1];
    // check the attribute size
    int expectedSize = getxattr(path, attribute, value, 0);
    if (expectedSize <= 0) {
        return "";
    }
    // Allocates native memory in C.
    char *attributeValue = (char *)malloc((expectedSize + 1) * sizeof(char));
    // get the attribute
    int actualSize = getxattr(path, attribute, attributeValue, expectedSize);
    if (actualSize != expectedSize) {
        free(attributeValue);
        return "";
    }
    attributeValue[expectedSize] = '\0';
    return attributeValue;
}

// free a string
void free_string(char *str)
{
    // Free native memory in C which was allocated in C.
    free(str);
}

#endif

File: lib/flutter_xattr.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'flutter_xattr_bindings_generated.dart' as bindings;

class Xattr {
  // get an xattr attribute on a file.
  static String getFileAttribute(String path, String attribute) {
    final filePathUtf8 = path.toNativeUtf8().cast<Char>();
    final attributeUtf8 = attribute.toNativeUtf8().cast<Char>();
    final attributeValueUtf8 = bindings.get_attribute(
      filePathUtf8,
      attributeUtf8,
    );
    final attributeValue = attributeValueUtf8.cast<Utf8>().toDartString();
    calloc.free(filePathUtf8);
    calloc.free(attributeUtf8);
    if (attributeValue.isNotEmpty) {
      bindings.free_string(attributeValueUtf8);
    }
    return attributeValue;
  }
}
